File: 5. Longest Palindromic Substring/longest-palindromic-substring.js
function longestPalindrome(s) {
    if (!s || s.length < 1) return '';

    function expandAroundCenter(left, right) {
        while (left >= 0 && right < s.length && s[left] === s[right]) {
            left--;
            right++;
        }
        return s.substring(left + 1, right);
    }

    let longest = '';

    for (let i = 0; i < s.length; i++) {
        const p1 = expandAroundCenter(i, i);
        const p2 = expandAroundCenter(i, i + 1);

        if (p1.length > longest.length) longest = p1;
        if (p2.length > longest.length) longest = p2;
    }

    return longest;
}

function main() {
    console.log(longestPalindrome("babad")); // "bab" o "aba"
    console.log(longestPalindrome("cbbd"));  // "bb"
    console.log(longestPalindrome("a"));     // "a"
    console.log(longestPalindrome("ac"));    // "a" o "c"
}

main();

File: 005. Longest Palindromic Substring/longest-palindromic-substring.py
def longestPalindrome(s):
    if not s:
        return ""

    def expandAroundCenter(left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left+1:right]

    longest = ""

    for i in range(len(s)):
        p1 = expandAroundCenter(i, i)
        p2 = expandAroundCenter(i, i+1)
        
        if len(p1) > len(longest):
            longest = p1
        if len(p2) > len(longest):
            longest = p2

    return longest

if __name__ == "__main__":
    print(longestPalindrome("babad"))  # "bab" o "aba"
    print(longestPalindrome("cbbd"))   # "bb"
    print(longestPalindrome("a"))      # "a"
    print(longestPalindrome("ac"))     # "a" o "c"
